File: convex/weightHistory.ts
import { query, mutation } from "./_generated/server";
import { v } from "convex/values";

// Save weight history (matches entity/weight_history.rs)
export const save = mutation({
  args: {
    medicationWeight: v.number(),
    dosageWeight: v.number(),
    quantityWeight: v.number(),
    priceWeight: v.number(),
    recencyWeight: v.number(),
    source: v.string(),
    sampleCount: v.number(),
  },
  handler: async (ctx, args) => {
    return await ctx.db.insert("weightHistory", {
      ...args,
      createdAt: Date.now(),
    });
  },
});

// Get current (most recent) weights
export const getCurrent = query({
  args: {},
  handler: async (ctx) => {
    return await ctx.db
      .query("weightHistory")
      .withIndex("by_created")
      .order("desc")
      .first();
  },
});

// Get weight history
export const getHistory = query({
  args: { limit: v.optional(v.number()) },
  handler: async (ctx, args) => {
    return await ctx.db
      .query("weightHistory")
      .withIndex("by_created")
      .order("desc")
      .take(args.limit ?? 50);
  },
});

// Get by ID
export const getById = query({
  args: { id: v.string() },
  handler: async (ctx, args) => {
    const items = await ctx.db.query("weightHistory").collect();
    return items.find((i) => i._id.toString() === args.id) ?? null;
  },
});

// Count all weight history entries
export const count = query({
  args: {},
  handler: async (ctx) => {
    const all = await ctx.db.query("weightHistory").collect();
    return all.length;
  },
});

File: core/crates/db/src/convex/weight_history.rs
//! Convex WeightHistoryRepository implementation

use std::sync::Arc;

use async_trait::async_trait;

use super::client::ConvexClient;
use crate::Result;
use crate::convex_args;
use crate::traits::{WeightHistoryModel, WeightHistoryRepository};

/// Convex-backed weight history repository
pub struct ConvexWeightHistoryRepo {
    client: Arc<ConvexClient>,
}

impl ConvexWeightHistoryRepo {
    pub fn new(client: Arc<ConvexClient>) -> Self {
        Self { client }
    }
}

#[async_trait]
impl WeightHistoryRepository for ConvexWeightHistoryRepo {
    async fn save(&self, history: &WeightHistoryModel) -> Result<WeightHistoryModel> {
        // WeightHistoryModel: id (Uuid), medication_weight, dosage_weight, quantity_weight, price_weight, recency_weight, source, sample_count, created_at
        let _id: String = self
            .client
            .mutation(
                "weightHistory:save",
                convex_args! {
                    "medicationWeight" => history.medication_weight,
                    "dosageWeight" => history.dosage_weight,
                    "quantityWeight" => history.quantity_weight,
                    "priceWeight" => history.price_weight,
                    "recencyWeight" => history.recency_weight,
                    "source" => &history.source,
                    "sampleCount" => history.sample_count
                },
            )
            .await?;

        Ok(history.clone())
    }

    async fn get_current(&self) -> Result<Option<WeightHistoryModel>> {
        self.client
            .query("weightHistory:getCurrent", convex_args!())
            .await
    }

    async fn get_history(&self, limit: i64) -> Result<Vec<WeightHistoryModel>> {
        self.client
            .query(
                "weightHistory:getHistory",
                convex_args! { "limit" => limit },
            )
            .await
    }

    async fn get_by_id(&self, id: &str) -> Result<Option<WeightHistoryModel>> {
        self.client
            .query("weightHistory:getById", convex_args! { "id" => id })
            .await
    }

    async fn count(&self) -> Result<i64> {
        self.client
            .query("weightHistory:count", convex_args!())
            .await
    }
}
